File: golang-for-nodejs-developers/examples/tcp_server.js
import net from 'node:net';

function handler(socket) {
  socket.write('Received: ');
  socket.pipe(socket);
}

const server = net.createServer(handler);
server.listen(3000, () => {
  console.log('TCP server listening on port 3000');
  // Auto-close for verification purposes if needed, 
  // but usually we just want to see it starts.
});

File: golang-for-nodejs-developers/examples/tcp_server.go
package main

import (
	"bufio"
	"fmt"
	"net"
)

func handler(conn net.Conn) {
	defer conn.Close()
	reader := bufio.NewReader(conn)

	for {
		message, err := reader.ReadString('\n')
		if err != nil {
			return
		}

		conn.Write([]byte("Received: "))
		conn.Write([]byte(message))
	}
}

func main() {
	listener, err := net.Listen("tcp", ":3000")
	if err != nil {
		panic(err)
	}

	defer listener.Close()
	fmt.Println("TCP server listening on port 3000")

	for {
		conn, err := listener.Accept()
		if err != nil {
			fmt.Println("Error accepting connection:", err)
			continue
		}

		go handler(conn)
	}
}
